udata/auth/views.py

```diff
@@ -19,6 +32,91 @@ def slash_url_suffix(url, suffix):
     return url.endswith('/') and ('%s/' % suffix) or ('/%s' % suffix)
 
 
+def send_change_email_confirmation_instructions(user, new_email):
+    data = [str(current_user.fs_uniquifier), hash_data(current_user.email), new_email]
+    token = _security.confirm_serializer.dumps(data)
+    confirmation_link = url_for('security.confirm_change_email', token=token, _external=True)
+
+    subject = _('Confirm change of email instructions')
+    send_mail(
+        subject=subject,
+        recipient=new_email,
+        template='confirmation_instructions',
+        user=current_user,
+        confirmation_link=confirmation_link
+    )
+
+
+def confirm_change_email_token_status(token):
+    expired, invalid, user, token_data = get_token_status(
+        token, 'confirm', 'CONFIRM_EMAIL', return_data=True)
+    new_email = None
+
+    if not invalid and user:
+        user_id, token_email_hash, new_email = token_data
+        invalid = not verify_hash(token_email_hash, user.email)
+
+    return expired, invalid, user, new_email
+
+
+def confirm_change_email(token):
+    expired, invalid, user, new_email = (
+        confirm_change_email_token_status(token))
+
+    if not user or invalid:
+        invalid = True
+        do_flash(*get_message('INVALID_CONFIRMATION_TOKEN'))
+    if expired:
+        send_change_email_confirmation_instructions(user, new_email)
+        do_flash(*(
+            (
+                'You did not confirm your change of email within {0}. '
+                'New instructions to confirm your change of email have '
+                'been sent to {1}.').format(
+                    _security.confirm_email_within, new_email),
+            'error'))
+    if invalid or expired:
+        return redirect(url_for('site.home'))
+
+    if user != current_user:
+        logout_user()
+        login_user(user)
+
+    if user.email == new_email:
+        msg = (
+            'Your change of email has already been confirmed.'
+            'info')
+    else:
+        user.email = new_email
+        _datastore.put(user)
+        msg = (
```

Comment:
I'm wondering if we should we send a success email as well? (similar to what is done when you reset your password?)